htlcswitch/switch.go

```diff
@@ -350,36 +350,89 @@ func (s *Switch) ProcessContractResolution(msg contractcourt.ResolutionMsg) erro
 }
 
 // SendHTLC is used by other subsystems which aren't belong to htlc switch
-// package in order to send the htlc update.
+// package in order to send the htlc update. This method is asynchronous, and
+// in case of a successful HTLC settle the preimage will be delivered on the
+// returned channel. In case an error is encountered during forwarding, it will
+// be returned on the error channel.
 func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID,
 	htlc *lnwire.UpdateAddHTLC,
-	deobfuscator ErrorDecrypter) ([sha256.Size]byte, error) {
+	deobfuscator ErrorDecrypter) (chan [sha256.Size]byte, chan error) {
+
+	preimageChan := make(chan [sha256.Size]byte, 1)
+	errChan := make(chan error, 1)
+
+	// We get a new unique payment ID that we'll potentially use for this
+	// HTLC, if it is not already in flight.
+	newPaymentID, err := s.paymentSequencer.NextID()
+	if err != nil {
+		errChan <- err
+		return preimageChan, errChan
+	}
 
 	// Before sending, double check that we don't already have 1) an
 	// in-flight payment to this payment hash, or 2) a complete payment for
 	// the same hash.
-	if err := s.control.ClearForTakeoff(htlc); err != nil {
-		return zeroPreimage, err
-	}
+	s.pendingMutex.Lock()
+	paymentID, preimage, err := s.control.ClearForTakeoff(newPaymentID, htlc)
+	switch {
 
-	// Create payment and add to the map of payment in order later to be
-	// able to retrieve it and return response to the user.
-	payment := &pendingPayment{
-		err:          make(chan error, 1),
-		preimage:     make(chan [sha256.Size]byte, 1),
-		paymentHash:  htlc.PaymentHash,
-		amount:       htlc.Amount,
-		deobfuscator: deobfuscator,
-	}
+	// If the payment was either not yet initiated, or already in flight,
+	// we'll add it to the set of pending payments, such that we can
+	// respond to the caller when the result comes back.
+	case err == nil || err == ErrPaymentInFlight:
+		// Create payment and add to the map of payment in order later to be
+		// able to retrieve it and return response to the user.
+		payment := &pendingPayment{
+			err:          errChan,
+			preimage:     preimageChan,
+			paymentHash:  htlc.PaymentHash,
+			amount:       htlc.Amount,
+			deobfuscator: deobfuscator,
+		}
+		s.pendingPayments[paymentID] = payment
```

Comment:
what if this overwrites an existing entry? seems the prior caller would never receive anything on their channels. do we need to support multiple subscribers?

Reply:
The thought here was only to handle the restart case, and otherwise make the caller responsible for not sending the HTLC more than once. But I definitely see the advantages of just handling duplicate entries, will see if can be easily handled!

Reply:
> and otherwise make the caller responsible for not sending the HTLC more than once

Isn't that hard to do if they have to call `SendHTLC` to see if they've already sent/re-register for ntfns?

Can you describe the scenario we're trying to solve?

Reply:
Went back to return results to all callers, so no overwriting would happen!